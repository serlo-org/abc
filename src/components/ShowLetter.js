import { Audio } from 'expo';
import React from 'react';
import { View, Text } from 'react-native';

import speakerImage from '../assets/images/speaker.png';
import repeatIcon from '../assets/images/repeat.png';
import repeatSource from '../assets/sounds/repeat.mp3';

Audio.setIsEnabledAsync(true);

import { RoundButton, IconWithBackground } from '../components/Components';

const styles = {
  container: {
    flex: 1,
    backgroundColor: '#00B4D5',
    alignItems: 'center',
    justifyContent: 'space-around'
  },
  header: {
    flexDirection: 'row',
    alignItems: 'flex-end'
  },
  bigLetter: {
    paddingLeft: 50,
    height: 200,
    width: 200,
    color: '#FFF',
    fontSize: 180,
    fontWeight: 'bold',
    textAlign: 'center'
  }
};

const ShowLetter = ({ letter, sound, isRepeat }) => {
  const letterSound = new Audio.Sound({ source: sound });
  const repeatSound = new Audio.Sound({ source: repeatSource });
  letterSound.loadAsync();
  repeatSound.loadAsync();

  const play = () => {
    if (isRepeat) this.icon.unfocus();
    letterSound.playAsync();
    letterSound.setPlaybackFinishedCallback(() => {
      letterSound.setPositionAsync(0);
      if (isRepeat) this.icon.focus();
      repeatSound.playAsync();
      repeatSound.setPlaybackFinishedCallback(() => {
        repeatSound.setPositionAsync(0);
      });
    });
  };

  const toggleRepeatButton = () => {
    if (isRepeat) {
      return (
        <IconWithBackground
          ref={view => {
            this.icon = view;
          }}
          icon={repeatIcon}
          size={40}
        />
      );
    }

    return null;
  };

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.bigLetter}>
          {letter}
        </Text>
        <RoundButton icon={speakerImage} size={40} onPress={play} />
      </View>
      <View height={80}>
        {toggleRepeatButton()}
      </View>
    </View>
  );
};

export default ShowLetter;
