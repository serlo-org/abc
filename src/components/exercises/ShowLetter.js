import React from 'react';
import { View, Text } from 'react-native';

import repeatIcon from '../../assets/images/repeat.png';
import speakerImage from '../../assets/images/speaker.png';
import { WHITE, PRIMARY } from '../../styles/colors';
import { DEFAULT } from '../../styles/text';
import { loadSounds } from '../helpers/audio';
import { RoundButton, IconWithBackground } from '../Components';

const styles = {
  container: {
    flex: 1,
    backgroundColor: PRIMARY,
    alignItems: 'center',
    justifyContent: 'space-around'
  },
  header: {
    flexDirection: 'row',
    alignItems: 'flex-end'
  },
  bigLetter: {
    ...DEFAULT,
    paddingLeft: 50,
    fontSize: 180
  }
};

const UnwrappedShowLetter = ({
  letter,
  sounds: [letterSound, repeatSound],
  isRepeat
}) => {
  const play = () => {
    if (isRepeat) this.icon.unfocus();
    letterSound.playAsync();
    letterSound.setPlaybackFinishedCallback(() => {
      letterSound.setPositionAsync(0);
      if (isRepeat) {
        this.icon.focus();
        repeatSound.playAsync();
        repeatSound.setPlaybackFinishedCallback(() => {
          repeatSound.setPositionAsync(0);
        });
      }
    });
  };

  const toggleRepeatButton = () => {
    if (isRepeat) {
      return (
        <IconWithBackground
          ref={view => {
            this.icon = view;
          }}
          icon={repeatIcon}
          size={40}
        />
      );
    }

    return null;
  };

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.bigLetter}>
          {letter}
        </Text>
        <RoundButton icon={speakerImage} size={40} onPress={play} />
      </View>
      <View height={80}>
        {toggleRepeatButton()}
      </View>
    </View>
  );
};

const WrappedShowLetter = loadSounds(UnwrappedShowLetter);

const ShowLetter = ({ sound, ...props }) => {
  const sounds = props.isRepeat
    ? [sound, require('../../assets/sounds/repeat.mp3')]
    : [sound];
  return <WrappedShowLetter sounds={sounds} {...props} />;
};

export default ShowLetter;
