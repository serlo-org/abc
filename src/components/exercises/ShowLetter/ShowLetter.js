import { Ionicons } from '@expo/vector-icons';
import React from 'react';
import { View, Text } from 'react-native';

import repeatIcon from '../../../assets/images/repeat.png';
import { playAll } from '../../../helpers/audio';
import { DEFAULT } from '../../../styles/text';
import IconWithBackground from '../../common/IconWithBackground';
import RoundButton from '../../common/RoundButton';
import { LoadSounds } from '../../helpers/Audio';

const styles = {
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-around'
  },
  header: {
    flexDirection: 'row',
    alignItems: 'flex-end'
  },
  bigLetter: {
    ...DEFAULT,
    paddingLeft: 50,
    fontSize: 180
  }
};

const UnwrappedShowLetter = ({ letter, sounds, isRepeat }) => {
  const play = () => {
    if (isRepeat) this.icon.unfocus();
    playAll(sounds).then(() => {
      if (isRepeat) this.icon.focus();
    });
  };

  const toggleRepeatButton = () => {
    if (isRepeat) {
      return (
        <IconWithBackground
          ref={view => {
            this.icon = view;
          }}
          icon={repeatIcon}
          size={40}
        />
      );
    }

    return null;
  };

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.bigLetter}>{letter}</Text>
        <RoundButton
          IconComponent={Ionicons}
          name="md-volume-up"
          size={40}
          onPress={play}
        />
      </View>
      <View height={80}>{toggleRepeatButton()}</View>
    </View>
  );
};

const ShowLetter = ({ sound, ...props }) => {
  const sounds = props.isRepeat
    ? [sound, require('../../../assets/sounds/repeat.mp3')]
    : [sound];

  return (
    <LoadSounds
      sounds={sounds}
      render={sounds => {
        return <UnwrappedShowLetter {...props} sounds={sounds} />;
      }}
    />
  );
};

export default ShowLetter;
