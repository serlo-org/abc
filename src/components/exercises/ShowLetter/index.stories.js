import { storiesOf } from '@storybook/react-native';

import * as ShowLetter from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/ShowLetter', module);

createStoriesFromFixtures(story, ShowLetter);
