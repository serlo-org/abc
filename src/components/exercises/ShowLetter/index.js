import { InfoScreen } from '../../../exercises';
import Component from './ShowLetter';

export const Exercise = InfoScreen.Exercise;
export const fixtures = InfoScreen.createFixtures([
  {
    name: 'A',
    props: {
      letter: 'A',
      sound: require('../../../assets/sounds/a.mp3')
    }
  },
  {
    name: 'a',
    props: {
      letter: 'a',
      sound: require('../../../assets/sounds/a.mp3')
    }
  }
]);
export { Component };
