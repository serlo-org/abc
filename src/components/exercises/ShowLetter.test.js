import React from 'react';
import renderer from 'react-test-renderer';

import ShowLetter from './ShowLetter';

it('renders without crashing (without repeat)', () => {
  const tree = renderer.create(
    <ShowLetter letter="A" sound={require('../../assets/sounds/a.mp3')} />
  );
  expect(tree).toMatchSnapshot();
});

it('renders without crashing (with repeat)', () => {
  const tree = renderer.create(
    <ShowLetter
      letter="B"
      sound={require('../../assets/sounds/b.mp3')}
      isRepeat
    />
  );
  expect(tree).toMatchSnapshot();
});
