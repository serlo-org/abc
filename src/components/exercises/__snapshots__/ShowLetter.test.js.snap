// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing (with repeat) 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00B4D5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFF",
          "fontSize": 180,
          "fontWeight": "bold",
          "height": 200,
          "paddingLeft": 50,
          "textAlign": "center",
          "width": 200,
        }
      }
    >
      B
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73DBFF",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    height={80}
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 9999,
          "padding": 20,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 40,
            "resizeMode": "contain",
            "width": 40,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`renders without crashing (without repeat) 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00B4D5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFF",
          "fontSize": 180,
          "fontWeight": "bold",
          "height": 200,
          "paddingLeft": 50,
          "textAlign": "center",
          "width": 200,
        }
      }
    >
      A
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73DBFF",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    height={80}
  />
</View>
`;
